libs/init/sbt/src/mill/main/sbt/SbtBuildGenMain.scala

```diff
@@ -1,542 +1,216 @@
+    val args0 = summon[ParserForClass[SbtBuildGenArgs]].constructOrExit(args.toSeq)
+    import args0.*
 
+    val sbtCmd = customSbt.getOrElse(defaultSbt.fold(sys.error, identity))
+    addExportPlugin()
+    val exportDir = os.temp.dir()
     try {
-      if (isWindows) {
-        /*
-        `-addPluginSbtFile` somehow doesn't work on Windows, therefore, the ".sbt" file is put directly in the `sbt` "project" directory.
-        The error message:
-        ```text
-        [error] Expected ':'
-        [error] Expected '='
-        [error] Expected whitespace character
-        [error] -addPluginSbtFile
-        [error]                  ^
-        ```
-         */
-        val sbtFile = writeTempSbtFileInSbtProjectDirectory(workspace)
-        os.call(
-          (sbtExecutable, "millInitExportBuild"),
-          cwd = workspace,
-          stdout = os.Inherit
-        )
-        os.remove(sbtFile)
-      } else
-        os.call(
-          (sbtExecutable, s"-addPluginSbtFile=${writeSbtFile().toString}", "millInitExportBuild"),
-          cwd = workspace,
-          stdout = os.Inherit
-        )
-
+      // Run task with cross-build prefix to export data for all cross Scala versions.
+      os.proc(sbtCmd, s"-DmillInitExportDir=$exportDir", "+millInitExportBuild")
+        .call(stdout = os.Inherit)
     } catch {
       case e: os.SubprocessException => throw RuntimeException(
           "The sbt command to run the `millInitExportBuild` sbt task has failed, " +
             s"please check out the following solutions and try again:\n" +
             s"1. check whether your existing sbt build works properly;\n" +
             s"2. make sure there are no other sbt processes running;\n" +
             s"3. clear your build output and cache;\n" +
-            s"4. update the project's sbt version to the latest or our tested version v${Versions.sbtVersion};\n" +
+            s"4. update the project's sbt version to the latest or our tested version v$sbtVersion;\n" +
             "5. check whether you have the appropriate Java version.\n",
           e
         )
-      case t: Throwable => throw t
     }
 
-    val buildExportPickled = os.read(workspace / "target" / "mill-init-build-export.json")
-    // TODO This is mainly for debugging purposes. Comment out or uncomment this line as needed.
-    // println("sbt build export retrieved: " + buildExportPickled)
-    import upickle.*
-    val buildExport = read[BuildExport](buildExportPickled)
+    // For cross projects, multiple specs are exported with settings for each Scala version.
+    type ExportedPackageSpec = (
+        // A flag that is set for cross-platform projects and the package location.
+        // When the flag is set, the location is the shared root directory for member projects.
+        // Otherwise, the location is the project base directory.
+        (Boolean, Seq[String]),
+        ModuleSpec
+    )
```

Comment:
Please use named tuples here, so we better see what we're dealing with in the use sites.

Reply:
Having two separate upickle setups for the same format is a call for trouble. It's not obvious and might break late after future refactorings.

I'd prefer, if we code-share the serialization format between the sbt plugin and this Mill implementation. Defining a case class in `api` module looks like the best approach.